### DISCUSSION TITLE
pyodbc is consuming 90% of my api call time. How to optimise it?

### DISCUSSION THREAD

**Discussion by AjayEdupuganti** [mkleehammer/pyodbc#1349] [Q&A]
My api call time is 6seconds and this pyodbc consuming 5.5 seconds. Out of which majorly is being consumed by fetchall in cursor.
How can i optimise this?
![image](https://github.com/mkleehammer/pyodbc/assets/140388245/f947203f-fd74-4d6d-801b-22282cd453ea)
![image](https://github.com/mkleehammer/pyodbc/assets/140388245/43d322de-0b3e-4ef9-9ea0-ee8e3205b16a)

**Comment by v-chojas** [comment]
It's unlikely to be pyODBC itself, and probably waiting for the DB server. If only you'd provide more information about your environment, maybe then we might have a better idea what's going on...

**Reply by AJAYKE** [reply]
You can see the connection time.
It's a flask server.
using Mssql.
Do let me know what other information do you need to help me with this.

**Reply by gordthompson** [reply]
Is `fetchall()` pulling rows over a network connection? If so, is it a local network (LAN) or an Internet connection (cloud)?

BTW, as @v-chojas suggests, the times reported by cProfile include the whole time that `fetchall()` is processing the rows, including the time required to receive those rows from the database (network latency). When I run a single select of 10_000 rows …

```python
rows = crsr.execute("SELECT TOP 10000 * FROM million_rows").fetchall()
```

… cProfile reports

```
         290 function calls (289 primitive calls) in 0.803 seconds

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.443    0.443    0.443    0.443 {method 'fetchall' of 'pyodbc.Cursor' objects}
```

while a single select of 100_000 rows …

```python
rows = crsr.execute("SELECT TOP 100000 * FROM million_rows").fetchall()
```

… produces

```
         290 function calls (289 primitive calls) in 4.078 seconds

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    3.778    3.778    3.778    3.778 {method 'fetchall' of 'pyodbc.Cursor' objects}
```

**Reply by AJAYKE** [reply]
you are saying 450ms to fetch 10k records is the python's best?
@gordthompson  it is pulling rows from the cloud. I understand that it fetches the rows, and creates a named tuple for each row which Is why it is taking time. it is showing the entire time it is taking ryt.
But is there a way to optimise it? Or is this how it is? 

https://www.progress.com/tutorials/odbc/performance-optimized
I found this article but didn't understand much, and couldn't find all the driver API functions on pyodbc documentation.

**Reply by gordthompson** [reply]
> you are saying 450ms to fetch 10k records is the python's best?

No, that's just the result I got while testing on **my** database over the LAN in **my** office. Other environments will produce different results.

> But is there a way to optimise it? Or is this how it is?

As far as `.fetchall()` itself is concerned, that's just how it is.

**Reply by AJAYKE** [reply]
so you are saying vertical scaling will improve the performance of the db calls.

**Reply by v-chojas** [reply]
In any case unless you can show the bottleneck happening in pyODBC code and not elsewhere, this is not the fault of pyODBC.

**Reply by AJAYKE** [reply]
@v-chojas so you are saying, the bottleneck could be because of poor network connection and might not be cursor.fetchall()

**Reply by v-chojas** [reply]
That's for you to figure out.
